docs/proposals/20201019-prometheus-rules-in-observatorium-api.md

```diff
@@ -0,0 +1,94 @@
+
+## Why
+
+The single biggest source of frustration from internal Red Hat users of Observatorium right now is the time it takes for changes to their rules to appear in our production infrastructure.
+
+### Pitfalls of the current solution
+
+#### Implicit deploy cycle dependency
+
+Currently, the only method of modifying Prometheus rules in Observatorium is via the jsonnet definition in our repository. This implicitly ties an update of the rule configuration to a rollout of the entire infrastructure. This implicitly ties the deploy-cycle of Prometheus rules to the deploy-cycle of Observatorium itself. If a new set of rules needs to be deployed, our team is required to roll-out our production infrastructure.
+
+As the number of tenants we support and users we serve increases, the team responsible for the Observatorium installation is required to roll out production more frequently in order to satisfy user requests. This becomes an increasing impediment to user experience as the size of Observatorium increases.
+
+#### Shared configuration repositories
+
+For rules to be rolled out, users need the ability to raise pull requests against the Observatorium repositories that contain our source configuration. This works fine when the two teams co-exist within the same organisation, but it becomes a blocking issue when the two teams share a much more restricted (e.g. inter-company) trust boundary.
+
+This is not a blocker for Red Hat's internal offering, however this is not inline with Observatorium's stated goal (TODO: insert link) of offering a SaaS-like monitoring solution.
+
+## Goals
+
+* Users can update their Prometheus rules without intervention from the operating team.
+
+## Non-Goals
+
+* Rate-limiting or accounting of user resources.
+* Define or implement the underlying database technology.
+
+## How
+
+We propose to solve the above problem by implementing a multi-tenant API that allows users to create, read, update and delete Prometheus alerting and recording rules.
+
+Thanos Ruler operates on the rule configuration provided by users, this is consumed from a file in local file defined by Ruler's `--rule.file` flag. Therefore, this configuration that we ingest via the API needs to be replicated into the local data directory of the Thanos Ruler instances. This is achieved with the [thanos-ruler-syncer](https://github.com/observatorium/thanos-rule-syncer), an application that calls the API above and writes the results into a location that Thanos Rule can access.
+
+This solution must be backwards compatible with rule configuration currently defined in our repository. That is to say, to ease the transition from our current setup to this new API, Thanos Ruler should be able to operate on Rules both from the API and Supplied as configuration.
+
+### How should we persist user's rule configuration?
+
+In the first iteration of this API, we will persist the rule configuration to a local directory that is backed by a `PersistentVolumeClaim`.
```

Comment:
I don't think implementing this API by storing request data in files is a viable approach for Observatorium. Right now, the Observatorium API is run as a horizontally scalable service, meaning we can have N replicas all serving requests. If we allow users to write data to disk via the Observatorium API, then suddenly each replica of the service will have a different view of the rules since the data in each Pod's PersistentVolumeClaim will be different. Using files to implement an API backend only works for read-only APIs that can be configured statically during deployment. For writable APIs we really need to store the data in an HA location to which all replicas of the Observatorium API have access, e.g. a database.

Reply:
Ok - that makes sense. The ability to scale horizontally is then another requirement that we need to explicitly call out :+1:

Reply:
Resolving - added explicitly as a requirement.